### PR TITLE
Upgrade to Ubuntu Bionic, Xenomai 3.0.8

### PR DESCRIPTION
**Pull request by egordon**
Note that this includes changes made to the `devel` branch in `libbarrett` since 2015. It has been tested on HERB by continuously running gravcomp along with the corresponding branch of `barrett_ros_control`:
https://github.com/personalrobotics/barrett_ros_control/tree/upgrade/bionic

Current master branch has been tagged v1.2.5.

### CODE REVIEW

**Comment by brianhou** [comment]
Hmm, there's also a different repository which I think contains some of these changes already: https://github.com/personalrobotics/libbarrett_devel. I'm not entirely sure of what the distinction is (i.e. why those changes weren't already in this repository). Can you investigate? (It'd also be helpful to isolate the PR diff to the main changes.)

**Comment by brianhou** [comment]
@aditya-vk please review

**Comment by egordon** [comment]
@aditya-vk @brianhou Ready for re-review!

**Review by brianhou — CHANGES_REQUESTED** [review]
This is a long PR so I want to try and summarize the necessary changes. It seems like there are three main categories: Xenomai/real-time, C++, and autotensioning (#9).

- Xenomai: new names for things (`native` -> `alchemy`, etc.)
- C++: use `constexpr` instead of `const` when appropriate (introduced in C++11)
- C++: remove `throw(optional_type_list)` exception specifications (deprecated in C++11)

Questions for my own understanding:
- Why do we need to wrap `malloc` and `free`?

Suggested changes:
- #9 introduced some logging code. I don't think that this is necessary to have all the time. Can we only turn this on if we compile in debug mode or something? Or just delete it entirely.
- Doxyfile should be deleted.
- Formatting is super screwy (mixed spaces and tabs). It seems that the libbarrett convention is tabs, so we should stick to that (despite our usual convention of spaces).

I'll do a more detailed pass when these changes are addressed. Thanks for the really great start!

File: programs/zerocal.cpp [hunk 1]
```diff
@@ -220,7 +231,7 @@ class AdjustJointStep : public CalibrationStep {
 					LowLevelWam<DOF>& llw = wam.getLowLevelWam();
 
 					// Record actual joint position, not commanded joint position
-					zeroPos[j] = wam.getJointPositions()[j];
+					zeroPos[j] = wam.getJointPositions()[j] - calPos[j];
```
Comment (brianhou) [inline programs/zerocal.cpp:235]: We should add a comment that this change is actually from the `zerocal` branch upstream.

File: programs/autotension_left_arm.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+rm ~/.barrett &&
+ln -s /etc/barrett/config_herblaunch_leftarm/ ~/.barrett &&
+
+# Run the script
+./programs/bt-wam-autotension -a left -j -1
```
Comment (brianhou) [inline programs/autotension_left_arm.sh:7]: Should `-1` just be `1`? Why is it only doing `j1`?

File: programs/autotension.cpp [hunk 3]
```diff
@@ -245,8 +292,13 @@ void AutoTension<DOF>::init(ProductManager& pm, std::vector<int> args) {
 	jpSlack1.resize(DOF);
 	jpSlack2.resize(DOF);
 
+	// Commenting the following block of code: the stack is 
+	// going out of real-time control when trying to initialize hand.
+	// TODO (avk): Find out what the issue is.
```
Comment (brianhou) [inline programs/autotension.cpp:259]: Does this issue still happen with @egordon's patches on top of the original PR?

File: programs/autotension.cpp [hunk 4]
```diff
@@ -256,208 +308,289 @@ void AutoTension<DOF>::init(ProductManager& pm, std::vector<int> args) {
...
 	jpInitial[1] = wam.getJointPositions();
+	jpInitial[1][1] = 0.0; // Stretched forward
+	jpInitial[1][3] = jpStopHigh[3] - stopBuffer[3]; // Folded opposite side as HERB home
+
+	// Start Position: Engage the Tang
 	jpStart[1] = jpInitial[1];
-	jpStart[1][1] = jpStopHigh[1] - tangBuffer[1];
-	jpStart[1][2] = jpStopLow[2] + tangBuffer[2];
-	jpStart[1][3] = jpStopHigh[3] - stopBuffer[3];
+	if (armToAutotension == "left")
```
Comment (brianhou) [inline programs/autotension.cpp:308]: This indentation is particularly misleading.

File: programs/autotension_left_arm.sh (hunk 2 above)
Comment (brianhou) [reply programs/autotension_left_arm.sh:7]: Never mind, I figured it out.

**Comment by egordon** [comment]
@brianhou Answer to "Why do we need to wrap malloc and free?"

Unfortunately, xenomai, as built according to the documentation, appears to contain calls to `_real_malloc()` and `_real_free()` (which, btw, is doubly-weird since it should be written in C++). These won't link to anything unless we compile `libbarrett` with the `--wrap=malloc` option, which causes the linker to link those function calls to the vanilla `malloc()` and `free()`.

However, by adding this option, existing calls to `malloc()` and `free()` try to link to `wrap_malloc()` and `wrap_free()`, which didn't exist in the code base. So I added them.

**Comment by egordon** [comment]
@brianhou Re logging code: `Logger` appears to log to a file, rather than `stdout/err`. Is this such a big deal to have in production code? My understanding was that file-logging was acceptable (e.g. see `syslog`, the ROS logs) in production code.

File: programs/autotension.cpp (hunk 3 above)
Comment (egordon) [reply programs/autotension.cpp:259]: @aditya-vk We should test this.

Comment (aditya-vk) [reply programs/autotension.cpp:259]: Right. I can test this next weekend, unless you can do it before. In any case, it is important to note that the hand should be closed before the tensioning process begins.

File: programs/autotension.cpp (hunk 4 above)
Comment (egordon) [reply programs/autotension.cpp:308]: Fixed, added clang-format with tabs.

**Comment by egordon** [comment]
@brianhou @aditya-vk Formatting fixed using clang-format (with modified configuration to use width-4 tabs instead of spaces).

Ready for re-review!

File: include/barrett/bus/can_socket.h [hunk 5]
```diff
@@ -58,14 +58,14 @@ class CANSocket : public CommunicationsBus {
 	/** CANSocket() Constructors
 	 */
 	CANSocket();
-	CANSocket(int port) throw(std::runtime_error);
+	CANSocket(int port);
```
Comment (aditya-vk) [inline include/barrett/bus/can_socket.h]: Can we add `explicit` keyword here to disallow hidden bugs?

File: include/barrett/config.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#ifndef __LIBBARRETT_CONFIG_H
```
Comment (brianhou) [inline include/barrett/config.h]: Why is this file necessary? I think CMake rewrites `include/barrett/config.h.in` to produce this file.

**Review by aditya-vk — COMMENTED** [review]
Looks good to me! Only a couple of formatting issues, which I think @brianhou is on top of as I type this.

**Review by brianhou — CHANGES_REQUESTED** [review]
My main concern is that I'm not sure whether the logger is actually useful, so I don't see the point in introducing code. I think we threw it together when we were trying to debug the WAM by reading the raw CAN messages. In retrospect, I'm not really clear on what scenarios this would be useful for debugging with. (I don't think we really learned much from it last time, aside from the usual message parsing fun.) For most issues we face, we'll be a higher level where we don't need to dive into the CAN messages.

If we do decide to keep it, my secondary annoyance is that the current implementation clutters up the home directory. I'd rather it be written to something under `/var/log` (and that we delete the commented-out code in `can_socket-rtcan.cpp`).

File: include/barrett/config.h (hunk 6 above)
Comment (egordon) [reply include/barrett/config.h]: Interestingly, after deleting the file and then re-running cmake, the file did not re-generate.

This is in contrast to `cmake/barrett-config.cmake`, which did re-generate.

So, for whatever reason, it appears to be necessary.

File: include/barrett/bus/can_socket.h (hunk 5 above)
Comment (egordon) [reply include/barrett/bus/can_socket.h]: Done!

File: include/barrett/config.h (hunk 6 above)
Comment (brianhou) [reply include/barrett/config.h]: The `config.h` file wouldn't be regenerated in the source directory. [This](https://github.com/personalrobotics/libbarrett/blob/upgrade/xenomai3/CMakeLists.txt#L206) shows that it should be written to `${PROJECT_BINARY_DIR}/include/barrett/config.h`. In contrast, `cmake/barrett-config.cmake.in` is written to `${PROJECT_SOURCE_DIR}/cmake/barrett-config.cmake` so it looks like it regenerates.

The question to me is whether this compiles with nothing at `${PROJECT_SOURCE_DIR}/include/barrett/config.h`. I think [it should](https://github.com/personalrobotics/libbarrett/blob/upgrade/xenomai3/CMakeLists.txt#L79), but it'd be great to confirm on `herb0`. If that's the case, then I think we can just delete it.

File: src/bus/can_socket-rtcan.cpp [hunk 7]
```diff
@@ -60,13 +65,20 @@ struct can_handle {
 };
 }
 
+double printCycle = 0.1;
```
Comment (brianhou) [inline src/bus/can_socket-rtcan.cpp]: Let's delete these five lines.

File: src/bus/can_socket-rtcan.cpp [hunk 8]
```diff
@@ -34,16 +34,21 @@
 #include <stdexcept>
 #include <cstdio>
 #include <cstring>
+#include <sstream>
```
Comment (brianhou) [inline src/bus/can_socket-rtcan.cpp]: I believe the `sstream` and `iostream` includes might be unnecessary.

**Review by brianhou — CHANGES_REQUESTED** [review]
Just nits now, this looks great!

**Review by aditya-vk — APPROVED** [review]
LGTM! Thanks a lot @egordon

File: include/barrett/config.h (hunk 6 above)
Comment (egordon) [reply include/barrett/config.h]: config.h removed

**Review by brianhou — APPROVED** [review]
